src/hardware/disney.cpp

```diff
@@ -17,63 +17,61 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
-#include "dosbox.h"
+#include "disney.h"
 
 #include <cassert>
-#include <queue>
 
-#include "bit_view.h"
 #include "checks.h"
-#include "inout.h"
-#include "mixer.h"
-#include "parallel_port.h"
-#include "pic.h"
-#include "setup.h"
 
 CHECK_NARROWING();
 
-class Disney {
-public:
-	Disney(const io_port_t base_port, const std::string_view filter_pref);
-	~Disney();
+Disney::Disney()
+        : LptDac("DISNEY", ChannelFeature::Mono, initialize_at_mixer_rate)
+{
+	// Prime the FIFO with a single silent sample
+	fifo.emplace(data_reg);
+}
 
-private:
-	Disney()                          = delete;
-	Disney(const Disney &)            = delete;
-	Disney &operator=(const Disney &) = delete;
+void Disney::BindToPort(const io_port_t lpt_port)
+{
+	using namespace std::placeholders;
 
-	AudioFrame Render();
-	void RenderUpToNow();
-	void AudioCallback(const uint16_t requested_frames);
+	// Register port handlers for 8-bit IO
+	const auto write_data = std::bind(&Disney::WriteData, this, _1, _2, _3);
+	const auto read_status = std::bind(&Disney::ReadStatus, this, _1, _2);
+	const auto write_control = std::bind(&Disney::WriteControl, this, _1, _2, _3);
+	BindHandlers(lpt_port, write_data, read_status, write_control);
+	LOG_MSG("LPT_DAC: Initialized Disney Sound Source on LPT port %03xh", lpt_port);
+}
 
-	bool IsFifoFull() const;
-	uint8_t ReadStatus(const io_port_t, const io_width_t);
-	void WriteData(const io_port_t, const io_val_t value, const io_width_t);
-	void WriteControl(const io_port_t, const io_val_t value, const io_width_t);
+void Disney::ConfigureFilters(const FilterState state)
+{
+	assert(channel);
+	// Run the ZoH up-sampler at the higher mixer rate
+	const auto mixer_rate_hz = check_cast<uint16_t>(channel->GetSampleRate());
+	channel->ConfigureZeroOrderHoldUpsampler(mixer_rate_hz);
+	channel->EnableZeroOrderHoldUpsampler();
 
-	// The DSS is an LPT DAC with a 16-level FIFO running at 7kHz
-	static constexpr auto dss_7khz_rate    = 7000;
-	static constexpr auto ms_per_frame     = 1000.0 / dss_7khz_rate;
-	static constexpr uint8_t max_fifo_size = 16;
+	// Pull audio frames from the Disney DAC at 7 kHz
+	channel->SetSampleRate(dss_7khz_rate);
+	ms_per_frame = 1000.0 / dss_7khz_rate;
```

Comment:
minor: `const auto millis_in_second = 1000.0`

Reply:
Might add this up in a timer or the mixer header. It's a common pattern.

Reply:
Good idea.